xds/type/matcher/v3/domain.proto

```diff
@@ -0,0 +1,51 @@
+syntax = "proto3";
+
+package xds.type.matcher.v3;
+
+import "xds/annotations/v3/status.proto";
+import "xds/type/matcher/v3/matcher.proto";
+
+import "validate/validate.proto";
+
+option java_package = "com.github.xds.type.matcher.v3";
+option java_outer_classname = "ServerNameMatcherProto";
+option java_multiple_files = true;
+option go_package = "github.com/cncf/xds/go/xds/type/matcher/v3";
+
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: Server name matcher]
+
+// Matches a fully qualified server name against a set of possibly overlapping
+// wildcard domain names.
+message ServerNameMatcher {
+  // Specifies an optional list of wildcard domains and a match action.
+  //
+  // The server name will be matched against all wildcard domains starting from
+  // the longest suffix, i.e. ``www.example.com`` will be first matched against
+  // ``www.example.com``, then ``*.example.com``, then ``*.com``. Note that
+  // partial wildcards are not supported, and values like ``*w.example.com``
+  // are invalid.
+  message DomainMatcher {
+    // A non-empty set of wildcard domain names, e.g. ``www.example.com`` or
+    // ``*.com``.
+    repeated string domains = 1 [ (validate.rules).repeated = {min_items : 1} ];
+
+    // Match action to apply when the server name matches any of the wildcard
+    // domain names.
+    Matcher.OnMatch on_match = 2;
+
+    // Indicates whether this match option should be considered when there is a
```

Comment:
Are there actually cases in which we would *not* want the most-specific match?

Reply:
For example, let's say we want to have 3 cases:
* `mail.google.com` - most specific match, may have some additional condition in sub-tree.
* `*.google.com` - do some redirect, should match `mail.google.com` if the nested condition fails.
* `*` - drop.

Having one default case can be handled with `on_no_match`. Having two default cases requires two levels of `on_no_match` which means some conditions have to be duplicated twice.

Reply:
I think I see what you want here, but the way it's expressed isn't very intuitive.  Instead, how about calling this option something like `ignore_if_nested_matcher_fails`?  That way, you can set it to true on the entry for `mail.google.com`, which will indicate that that entry will be ignored if the nested matcher fails, in which case it will fall back to `*.google.com` as the next-most-specific match.

Reply:
The name followed the pattern in https://github.com/cncf/xds/blob/main/xds/type/matcher/v3/ip.proto#L48, but I agree it is not super intuitive. What I want is actually structured fallback, so the bit should be set on `*.google.com` instead.

I realize this matcher is probably a candidate for virtual host matching in Envoy. @snowp is there some prior art to express trie-matching for wildcard domains?

Reply:
Dropped this field until we need it.